### PR TITLE
Netsim compose table layout

### PR DESCRIPTION
**Pull request by islemaster**
out

Ran this design for the "Send Message" box by Baker and Brook and Sarah, and they decided they liked it better than [the one in the mockups](http://wiki.code.org/display/PROD/Internet+Simulator+Mocks).  So here's my implementation of it.

This first pass always displays all four formats; the final version will probably only show binary and one other format at any given time, but I figure the easiest way to handle that is to hook up the whole thing and then just hide rows/columns we don't want.

There will probably be further opportunity to remove duplication as I work on the "received messages" section, but we're not quite there yet.

### CODE REVIEW

**Comment by bcjordan** [comment]
Did you get a chance to test the layout in IE9?

File: apps/src/netsim/NetSimSendWidget.js [hunk 1]
```diff
@@ -139,107 +129,225 @@ var whitelistCharacters = function (whitelistRegex) {
...
+
+    shortNumberFields.forEach(function (fieldName) {
+      rowFields[fieldName] = tr.find('input.' + fieldName);
+      rowFields[fieldName].keypress(
+          whitelistCharacters(rowType.shortNumberAllowedCharacters));
+      rowFields[fieldName].keyup(
+          this.makeKeyupHandler(fieldName, rowType.shortNumberConversion));
+      rowFields[fieldName].blur(
+          this.makeBlurHandler(fieldName, rowType.shortNumberConversion));
+    }.bind(this));
```
Comment (bcjordan) [inline apps/src/netsim/NetSimSendWidget.js]: Any reason for using `.bind` this versus `.forEach(function (){}, this)` ?

**Comment by islemaster** [comment]
I haven't, and you're absolutely right that I need to.  I'll fire that up right now.

File: apps/src/netsim/NetSimSendWidget.js [hunk 2]
```diff
@@ -139,107 +129,225 @@ var whitelistCharacters = function (whitelistRegex) {
...
+      messageConversion: function (asciiString) {
+        return asciiToBinary(asciiString, 8);
+      }
+    }
+  ];
+
+  rowTypes.forEach(function (rowType) {
+    var tr = rootDiv.find('tr.' + rowType.typeName);
+    this[rowType.typeName + 'UI'] = {};
+    var rowFields = this[rowType.typeName + 'UI'];
```
Comment (bcjordan) [inline apps/src/netsim/NetSimSendWidget.js]: maybe extract variable `var rowUI(Name|ID|Key|something else) = rowType.typeName + 'UI';`

File: apps/src/netsim/NetSimSendWidget.js (hunk 1 above)
Comment (islemaster) [inline apps/src/netsim/NetSimSendWidget.js]: Uh... because I didn't realize it was an option?

File: apps/src/netsim/NetSimSendWidget.js [hunk 3]
```diff
@@ -139,107 +129,225 @@ var whitelistCharacters = function (whitelistRegex) {
...
 };
 
-NetSimSendWidget.prototype.rebuildPacketBinary_ = function () {
-  this.packetBinary = packetEncoder.createBinary({
-    toAddress: unsignedIntegerToBinaryString(this.toAddress_, 4),
-    fromAddress: unsignedIntegerToBinaryString(this.fromAddress_, 4),
-    payload: this.binaryPayloadTextbox_.val().replace(/[^01]/g, '')
+/**
+ * Update send widget display
+ * @param {*} [skipElement]
```
Comment (bcjordan) [inline apps/src/netsim/NetSimSendWidget.js]: type?

**Comment by breville** [comment]
Cool UI.  One thing we found repeatedly from the signin/signup pages was
that people forgot what each field stood for once they started typing and
lost the in-field caption, so we ended up adding external labels that
showed permanently.  Might be worth keeping an eye out for that in
usability testing.

File: apps/src/netsim/NetSimSendWidget.js [hunk 4]
```diff
@@ -139,107 +129,225 @@ var whitelistCharacters = function (whitelistRegex) {
...
+
+/**
+ * Produces a single binary string in the current packet format, based
+ * on the current state of the widget (content of its internal fields).
+ * @returns {string} - binary representation of packet
+ * @private
+ */
+NetSimSendWidget.prototype.getPacketBinary_ = function () {
+  var encoder = new PacketEncoder([
+    { key: 'toAddress', bits: 4 },
```
Comment (bcjordan) [inline apps/src/netsim/NetSimSendWidget.js]: Maybe extract variable `var (something) = 4`

File: apps/src/netsim/NetSimSendWidget.js [hunk 5]
```diff
@@ -139,107 +129,225 @@ var whitelistCharacters = function (whitelistRegex) {
...
+    { key: 'message', bits: Infinity }
+  ]);
+  return encoder.createBinary({
+    toAddress: intToBinary(this.toAddress, 4),
+    fromAddress: intToBinary(this.fromAddress, 4),
+    packetIndex: intToBinary(this.packetIndex, 4),
+    packetCount: intToBinary(this.packetCount, 4),
+    message: this.message
+  });
+};
```
Comment (bcjordan) [inline apps/src/netsim/NetSimSendWidget.js]: ^ `\n`

File: apps/src/netsim/dataConverters.js [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+};
+
+/**
+ * Takes a set of whitespace-separated numbers and pads the spacing between
+ * them to the width of the widest number, so that they line up when they
+ * wrap.
+ * @param {string} decimalString
+ * @returns {string} aligned decimal string
+ */
+exports.formatDecimal = function (decimalString) {
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: not sure if this is accurate, but maybe s/`format`/``gridLayout`?

File: apps/src/netsim/dataConverters.js [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+  if (decimalString.replace(/\D/g, '') === '') {
+    return '';
+  }
+
+  var numbers = exports.minifyDecimal(decimalString).split(/\s+/);
+
+  // Find the widest number
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: misread that as wildest :tada:

File: apps/src/netsim/dataConverters.js [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+  var mostDigits = numbers.reduce(function(prev, cur) {
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: Nice!

File: apps/src/netsim/dataConverters.js [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    if (cur.length > prev) {
+      return cur.length;
+    }
+    return prev;
+  }, 0);
+
+  return numbers.map(function (numString) {
+    // Left-pad each number with non-breaking spaces up to max width.
+    while (numString.length < mostDigits) {
+      numString = "\xA0" + numString;
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: Potentially extract e.g. `var nbspCharacter`

File: apps/src/netsim/dataConverters.js [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+ * @returns {string}
+ */
+exports.binaryToHex = function (binaryString) {
+  var nibble;
+  var nibbleWidth = 4;
+  var chars = [];
+  var uglyBinary = exports.minifyBinary(binaryString);
+  for (var i = 0; i < uglyBinary.length; i += nibbleWidth) {
+    // Right-pad nibble with zeroes
+    nibble = zeroPadRight(uglyBinary.substr(i, nibbleWidth), nibbleWidth);
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: maybe `currentNibble` (was partially expecting it to be returned)

File: apps/test/netsim/dataConverters.js [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+      assertEqual('01100011', asciiToBinary('c', 8));
+      assertEqual('01100100', asciiToBinary('d', 8));
+      assertEqual('01000001', asciiToBinary('A', 8));
+      assertEqual('01000010', asciiToBinary('B', 8));
+      assertEqual('01000011', asciiToBinary('C', 8));
+      assertEqual('01000100', asciiToBinary('D', 8));
+    });
+
+    it ("takes multiple characters", function () {
+      assertEqual('010010000110100100100001', asciiToBinary('Hi!', 8));
```
Comment (bcjordan) [inline apps/test/netsim/dataConverters.js]: :wave:

File: apps/test/netsim/dataConverters.js [hunk 12]
```diff
@@ -0,0 +1,484 @@
+var testUtils = require('../util/testUtils');
```
Comment (bcjordan) [inline apps/test/netsim/dataConverters.js]: Tests should reference more hilariously outlandish ASCII chars

**Comment by bcjordan** [comment]
LGTM! Nice testing.

You are now extremely prepared for every binary / ascii / integer conversion interview question out there

File: apps/test/netsim/dataConverters.js (hunk 12 above)
Comment (laurelfan) [inline apps/test/netsim/dataConverters.js]: ☃

Comment (islemaster) [inline apps/test/netsim/dataConverters.js]: Consider it done™.

**Comment by islemaster** [comment]
Hmm... either IE9 hates NetSim, or Browserstack hates connecting to my machine.  I'm definitely getting near-random line numbers in Javascript errors, and it seems to fail to pull down netsim.js at all half the time.  Investigating.

**Comment by islemaster** [comment]
Here's the UI running in Win7 IE9 via Browserstack.  Actually, I'm pleased with how consistent it looks.  I did have to fix a problem with NetSimLogger to get things running in IE9.

out

**Comment by islemaster** [comment]
@breville Regarding the textbox watermarks, I actually had a header column at one point to indicate which row was which format.  I tore it out and switched to this to save space; but it would be easy to put back in, and I don't think the watermark hurts to leave in as well.

A couple other things that may reclaim some space:
1. I'm considering condensing the two packet fields into one.  They should never be user-editable (though I've hooked them up in case we want to let people play 'misbehaving client') and we might reasonably display "1 / 1" or "0001 0001" within a single textbox, reclaiming a little horizontal space.
2. Eventually there will be a global "format" control that will hide some rows, so that the student only sees the binary and one other format.  Maybe harder to forget which is which then?

**Pull request merged**